### ISSUE TITLE
usize is masked to u32 still triggers  error: casting `usize` to `u32` may truncate the value on targets with 64-bit wide pointers

### ISSUE THREAD

**Issue by simonsso** [rust-lang/rust-clippy#16854]
### Summary

I think Clippy should be smart enough to realize the `usize` has been masked to a size which fits in a `u32` 
I tired example below with both 

Running clippy on pedantic mode. 

`clippy 0.1.94 (e408947bfd 2026-03-25)`

### Lint Name

cast_possible_truncation

### Reproducer

Running clippy on

```rust
  let remaining = (levels.len().saturating_sub(MAX_PAYLOAD_NUM) & u32::MAX as usize) as u32;
  let remaining = (levels.len().saturating_sub(MAX_PAYLOAD_NUM) & 0xFFFF_FFFF_usize) as u32;
```

I saw this happen:

```
error: casting `usize` to `u32` may truncate the value on targets with 64-bit wide pointers
   --> src/file.rs:215:25
    |
215 |         let remaining = (levels.len().saturating_sub(MAX_PAYLOAD_NUM) & u32::MAX as usize) as u32;
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: if this is intentional allow the lint with `#[allow(clippy::cast_possible_truncation)]` ...
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.94.0/index.html#cast_possible_truncation
help: ... or use `try_from` and handle the error accordingly
    |
215 -         let remaining = (levels.len().saturating_sub(MAX_PAYLOAD_NUM) & u32::MAX as usize) as u32;
215 +         let remaining = u32::try_from(levels.len().saturating_sub(MAX_PAYLOAD_NUM) & u32::MAX as usize);
    |
```

Replacing u32::MAX with 0xFFFF_FFFF yields same result:
```
  |
215 -         let remaining = (levels.len().saturating_sub(MAX_PAYLOAD_NUM) & 0xFFFF_FFFF_usize) as u32;
215 +         let remaining = u32::try_from(levels.len().saturating_sub(MAX_PAYLOAD_NUM) & 0xFFFF_FFFF_usize);
    |
```

I expected to see this happen:

Since the usize value is obviously masked to 32 bits it will always be within u32 range, there is nothing to worry about.

### Version

```text
Running clippy on pedantic mode. 
clippy 0.1.94 (e408947bfd 2026-03-25)
```

**Comment by Souradip121** [comment]
I've identified the root cause and have a fix ready.

Root cause: Two bugs in clippy_lints/src/casts/cast_possible_truncation.rs:

get_constant_bits() couldn't evaluate u32::MAX as usize because ConstEvalCtxt has no ExprKind::Cast handler — it returned None instead of Some(32). Fixed by peeling through cast layers to reach the underlying constant.

The usize → fixed-size int branch ignored the reduced bit count from apply_reductions() entirely — it checked to_nbits <= 32 unconditionally instead of also checking from_nbits > to_nbits. Fixed by adding that condition.

After both fixes, (x & u32::MAX as usize) as u32, (x & 0xFFFF_FFFF_usize) as u32, and (x.saturating_sub(n) & u32::MAX as usize) as u32 no longer trigger the lint. Unmasked x as u32 still lints correctly. All existing tests pass.

@rustbot claim

**Comment by simonsso** [comment]
@Souradip121  do you know why it is stuck in the review queue?
